### PR TITLE
antimatter-dimensions: init at 0-unstable-2024-05-11

### PR DESCRIPTION
**Pull request by amozeo**
## About package

Antimatter Dimensions is an idle incremental game with many prestige layers, and can take up to year to complete it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 317242` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>antimatter-dimensions</li>
  </ul>
</details>

File: pkgs/by-name/an/antimatter-dimensions/package.nix [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+, electron
+, fetchFromGitHub
+, lib
+, makeDesktopItem
+, writeScriptBin
+, electronAppName ? "Antimatter Dimensions"
+}:
+
+let
+ dummyElectronApp = ./app.js;
```
Comment (JohnRTitor) [inline pkgs/by-name/an/antimatter-dimensions/package.nix:11]: Why is this needed?

File: pkgs/by-name/an/antimatter-dimensions/package.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+in
+buildNpmPackage rec {
+  pname = "antimatter-dimensions";
+  version = "unstable-2024-05-11";
```
Comment (JohnRTitor) [inline pkgs/by-name/an/antimatter-dimensions/package.nix:15]:
```suggestion
  version = "0-unstable-2024-05-11";
```

Since there is no tagged version, we should hardcode 0.

File: pkgs/by-name/an/antimatter-dimensions/package.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      name = "antimatter-dimensions";
+      exec = "antimatter-dimensions";
+      icon = "antimatter-dimensions";
+      desktopName = electronAppName;
+      comment = meta.description;
+      categories = [ "Game" ];
+      terminal = false;
+    })
+  ];
+
```
Comment (JohnRTitor) [inline pkgs/by-name/an/antimatter-dimensions/package.nix:62]:
```suggestion

  passthru.updateScript = unstableGitUpdater {
    hardcodeZeroVersion = true;
  };

```

File: pkgs/by-name/an/antimatter-dimensions/package.nix (hunk 1 above)
Comment (amozeo) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:12]: build of the project doesn't provide app.js, and that file is used to change the directory where data is stored
if you know another way to change the electron application name, like with environment variables, I am happy to use it.

Comment (JohnRTitor) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:12]: Ok, let's make that clear in the comments, and please also mention that if in future this issue is resolved upstream, this workaround is to be removed.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Why bundle this as an electron app when you can have the desktop entry file run `xdg-open index.html`?

**Comment by JohnRTitor** [comment]
> Why bundle this as an electron app when you can have the desktop entry file run `xdg-open index.html`?

Let's not go there. I'd be very much against a scenario where I installed an app and it opens in the web browser to run for god's sake.

**Comment by amozeo** [comment]
> Why bundle this as an electron app when you can have the desktop entry file run `xdg-open index.html`?

[Steam](https://store.steampowered.com/app/1399720/Antimatter_Dimensions) version of upstream is bundled with electron, and it's not available for linux unless you use proton.

File: pkgs/by-name/an/antimatter-dimensions/package.nix [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+, makeDesktopItem
+, unstableGitUpdater
+, writeScriptBin
+, electronAppName ? "Antimatter Dimensions"
+}:
+
+let
+ dummyElectronApp = ./app.js;
+in
+buildNpmPackage rec {
```
Comment (Pandapip1) [inline pkgs/by-name/an/antimatter-dimensions/package.nix:14]: Maybe use `finalAttrs` instead of `rec`?

File: pkgs/by-name/an/antimatter-dimensions/package.nix [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    hash = "sha256-+G9mNilt5Ewja5P+Bt312EcCJknMu7FOMn5b4FseAyQ=";
+  };
+  nativeBuildInputs = [
+    # build script calls git to get git hash, message and author
+    # since fetchFromGitHub doesn't provide this information
+    # create a dummy git
+    (writeScriptBin "git" ''
+      echo "unknown"
+    '')
+  ];
```
Comment (Pandapip1) [inline pkgs/by-name/an/antimatter-dimensions/package.nix:30]:
```suggestion
  src = fetchFromGitHub {
    owner = "IvarK";
    repo = "AntimatterDimensionsSourceCode";
    rev = "b3a254af60207a03d04473bb81726e921f5b2c61";
    hash = "sha256-+G9mNilt5Ewja5P+Bt312EcCJknMu7FOMn5b4FseAyQ=";
    leaveDotGit = true;
  };
  nativeBuildInputs = [
    git
  ];
```

Obviously git has to also be added to the inputs.

File: pkgs/by-name/an/antimatter-dimensions/app.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+const { app, BrowserWindow } = require('electron');
```
Comment (Pandapip1) [inline pkgs/by-name/an/antimatter-dimensions/app.js:1]: Maybe it might make sense to use a more lightweight library such as https://www.npmjs.com/package/webview-nodejs?

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/an/antimatter-dimensions/package.nix (hunk 4 above)
Comment (amozeo) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:14]: tried `buildNpmPackage (self: {/*...*/})` however evaluation fails because buildNpmPackage doesn't support functions as argument.

File: pkgs/by-name/an/antimatter-dimensions/package.nix (hunk 5 above)
Comment (JohnRTitor) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:30]: using `leaveDotGit` will make the output non-deterministic, since obviously .git can change easily. You'll get random hash mismatchs. If you really need the commit hash, message, id, you can [design an updateScript](https://github.com/NixOS/nixpkgs/commit/7b136032a56ca0bf1eac8d32bbf298a8dafdf976), that'll generate a structured output (json) and use it.

File: pkgs/by-name/an/antimatter-dimensions/app.js (hunk 6 above)
Comment (amozeo) [reply pkgs/by-name/an/antimatter-dimensions/app.js:1]: nixpkgs doesn't have these node packages.

Comment (Pandapip1) [reply pkgs/by-name/an/antimatter-dimensions/app.js:1]: Fair enough. Feel free to mark this as resolved.

File: pkgs/by-name/an/antimatter-dimensions/package.nix (hunk 5 above)
Comment (Pandapip1) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:30]: > using leaveDotGit will make the output non-deterministic

How so? The only stuff that uses `.git` is https://github.com/IvarK/AntimatterDimensionsSourceCode/blob/master/build/post-build.js#L9-L13, which depend entirely on the `HEAD` commit, which is fixed in `fetchFromGitHub`.

Comment (JohnRTitor) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:30]: See https://github.com/NixOS/nixpkgs/issues/8567.

Comment (Pandapip1) [reply pkgs/by-name/an/antimatter-dimensions/package.nix:30]: Fair enough. Mark this as resolved too.

Comment (JohnRTitor) [reply pkgs/by-name/an/antimatter-dimensions/package.nix]: @amozeo please revert this change, using `leaveDotGit` is not desireable.

**Comment by JohnRTitor** [comment]
You might also want to add a desktop file like https://github.com/NixOS/nixpkgs/blob/a71e967ef3694799d0c418c98332f7ff4cc5f6af/pkgs/applications/editors/vscode/generic.nix#L125-L142

**Comment by amozeo** [comment]
> You might also want to add a desktop file like
@JohnRTitor

what's wrong with the way I currently made desktop icon?

**Comment by JohnRTitor** [comment]
I am not seeing a desktop file in `/nix/store/xjz48hc4q7ijcxwbyxc330wqhqyjjvfs-antimatter-dimensions-0-unstable-2024-05-11/share/applications` for some reason though.

**Review by JohnRTitor — APPROVED** [review]
LGTM
